fluidPage (
  sidebarLayout(
    sidebarPanel(
      selectizeInput("gene","Manual input for gene list",choices=NULL,selected=NULL,multiple=TRUE),
      checkboxGroupInput("group","TCF group",choices = c("High","Low","Normal"),selected = c("High","Low","Normal")),
      checkboxGroupInput("bor","BOR",choices = c("PD","PR","SD","Normal"),selected = c("PD","PR","SD","Normal")),
      checkboxGroupInput("diagnose","Diagnose",choices = c("Melanoma","Ovarian Cancer","Normal"),selected = c("Melanoma","Ovarian Cancer","Normal")),
      checkboxGroupInput("histology","Histological class",choices = c("Cutaneous","Uveal","Serous","Other","Normal"),selected = c("Cutaneous","Uveal","Serous","Other","Normal")),
      selectizeInput("cluster","Cell clusters",choices=NULL,selected=NULL,multiple=TRUE), width=2
    ),
    mainPanel(
      tabsetPanel(
        tabPanel("Mean gene expression Boxplots",dataTableOutput("Table1"),plotOutput("Boxplot1"),dataTableOutput("Stat1")),
        tabPanel("Mean gene expression Heatmap and Dotplot",
                 sidebarPanel(radioButtons("scaled","Heatmap, scale by",c("none","row"),selected=c("row")),
                              radioButtons("cluster_cols","Heatmap cluster columns",c("yes"=TRUE,"no"=FALSE),selected = c("yes"=TRUE))),
                 mainPanel(plotOutput("Heatmap"),plotOutput("DotPlot"))),
        tabPanel("Cell type frequencues",dataTableOutput("Table2"),plotOutput("Boxplot2"),dataTableOutput("Stat2")),
        tabPanel("scTCRseq Summary",
                 sidebarPanel(selectInput("div_ind","What diversity index to plot?",choices = c("inv.simpson","shannon","chao1","ACE","gini.simpson","norm.entropy"),selected="inv.simpson"),
                              selectInput("group_ind","What groups to compare?", choices = c("group","BOR","ORIGDIAG","HISTCLAS","SUBJID"),selected = "group"),
                              selectInput("cloneCall","How to call the clone?",choices = c("VDJC gene"="gene","CDR3 nucleotides"="nt","CDR3 aminoacids"="aa","VDJC gene+ CDR3 nucleotides"="strict"),selected = c("VDJC gene+ CDR3 nucleotides"="strict")),
                              selectInput("method","Method to calculate clonal overlap",choices = c("overlap","morisita","jaccard","cosine","raw"),selected="raw"),width=3),
                 mainPanel(strong("Clonal diversity for samples and groupings."),
                           fluidRow(column(6,plotOutput("Diversity_plot1")),
                                    column(6,plotOutput("Diversity_plot2"))),
                           strong("Homeostasis and Proportion."),
                           p("Homeostasis  - the space occupied by clone proportions.
                             Proportions - the relative clonal space occupied by the clones."),
                           fluidRow(column(6,plotOutput("Plot_Homeostasis")),
                                    column(6,plotOutput("Plot_Proportion"))),
                           strong("Unique Clones"),
                           p("This function quantifies unique clones."),
                           fluidRow(column(6,plotOutput("Plot_UniqueClones")),
                                    column(6,plotOutput("Plot_UniqueClones_scaled"))),
                           strong("Clonal overlap between samples"),
                           plotOutput("Plot_clonalOverlap"),
                           strong("Distribution of clonal length"),
                           p("Displays either the nucleotide or amino acid sequence length for CDR3, combined clone (both chains)"),
                           plotOutput("Plot_clonalLength"),width=9)),
        tabPanel("scTCRseq Tcells",
                 sidebarPanel(selectInput("group_tcell","What groups to compare?", choices = c("T cells populations + TCF group"="Cell_type2_group","T cells Subpopulations + TCF group"="Cell_type_group"),selected = c("T cells populations + TCF group"="Cell_type2_group")),
                              checkboxGroupInput("Tcell_pop","T cell populations",choices = c("CD4","CD8","MAIT"),selected = c("CD4","CD8")),
                              checkboxGroupInput("Tcell_subpop","T cell subpopulations",
                                                 choices = c("CD4 Naive","CD8 Naive","CD4 SCM","CD8 SCM",
                                                                                                    "CD4 CM","CD8 CM","CD4 EM", "CD8 EM","CD8 CTL","MAIT"),
                                                 selected = c("CD4 Naive","CD8 Naive","CD4 SCM","CD8 SCM",
                                                               "CD4 CM","CD8 CM","CD4 EM", "CD8 EM","CD8 CTL","MAIT")),
                              width=3),
                 mainPanel(strong("Homeostasis."),
                           p("Homeostasis  - the space occupied by clone proportions."),
                           plotOutput("Plot_Homeostasis_Tcells"),
                           dataTableOutput("Table_tcell"),
                           strong("Unique Clones"),
                           p("This function quantifies unique clones."),
                           plotOutput("Plot_UniqueClones_scaled_Tcell"),width=9)),
        tabPanel("scTCRseq VDJ genes", 
                 sidebarPanel(selectInput("group_vdj","What groups to compare?", choices = c("SUBJID","group","BOR","ORIGDIAG","HISTCLAS"),selected = "group"),
                              selectInput("chain","Chain to show",choices = c("TRA", "TRB"), selected = "TRB"),
                              selectInput("vdj_gene1","Chain to show",choices = c("V", "D", "J"), selected = "V"),
                              textInput("vdj_gene2","VDJ gene of the TCR to visualize",value="TRBV")),
                 mainPanel(strong("Examining the relative amino acid composition by position."),
                           p("Plots the proportion of amino acids along the residues of the CDR3 amino acid sequence."),
                           plotOutput("CDR3_AAsomposition"),
                           strong("Examining the VDJ gene usage across clones."),
                           p("Plots the proportion V or J genes used by grouping variables. This function only quantifies single gene loci for indicated chain."),
                           plotOutput("VJ_usage1"),
                           strong("Visualizing the distribution of gene usage."),
                           p("This function will allow for the visualizing the distribution of the any VDJ and C gene of the TCR. This function requires assumes two chains were used in defining clone, if not, it will default to the only chain present regardless of the chain parameter."),
                           plotOutput("VDJgene_plot1"),plotOutput("VDJgene_plot2"),dataTableOutput("VDJ_table")))
      ), width = 10
    )
  )
)